sql/core/src/main/scala/org/apache/spark/sql/WindowFunctionDefinition.scala

```diff
@@ -0,0 +1,341 @@
+   * the partition. For example
+   * {{{
+   *   df.over(partitionBy("k1").orderBy($"k2", $"k3"))
+   * }}}
+   * @group window_funcs
+   */
+  def orderBy(cols: Column*): WindowFunctionDefinition = {
+    val sortOrder: Seq[SortOrder] = cols.map { col =>
+      col.expr match {
+        case expr: SortOrder =>
+          expr
+        case expr: Expression =>
+          SortOrder(expr, Ascending)
+      }
+    }
+    new WindowFunctionDefinition(column, partitionSpec, sortOrder, frame)
+  }
+
+  /**
+   * Returns the current [[WindowFunctionDefinition]]. This is a dummy function,
+   * which makes the usage more like the SQL.
+   * For example:
+   * {{{
+   *   df.over(partitionBy("k1").orderBy($"k2", $"k3").range.between.preceding(1).and.currentRow)
+   * }}}
+   * @group window_funcs
+   */
+  def between: WindowFunctionDefinition = {
+    assert(this.frame.isInstanceOf[SpecifiedWindowFrame], "Should be a WindowFrame.")
+    new WindowFunctionDefinition(column, partitionSpec, orderSpec, frame, true)
+  }
+
+  /**
+   * Returns a new [[WindowFunctionDefinition]] indicate that we need to specify the
+   * upper bound.
+   * For example:
+   * {{{
+   *   df.over(partitionBy("k1").orderBy($"k2", $"k3").range.between.preceding(3).and.currentRow)
+   * }}}
+   * @group window_funcs
+   */
+  def and: WindowFunctionDefinition = {
+    new WindowFunctionDefinition(column, partitionSpec, orderSpec, frame, false)
+  }
+
+  /**
+   * Returns a new Ranged [[WindowFunctionDefinition]].
+   * For example:
+   * {{{
+   *   df.over(partitionBy("k1").orderBy($"k2", $"k3").range.between.preceding(3).and.currentRow)
+   * }}}
+   * @group window_funcs
+   */
+  def range: WindowFunctionDefinition = {
+    new WindowFunctionDefinition(column, partitionSpec, orderSpec,
+      SpecifiedWindowFrame(RangeFrame, UnboundedPreceding, UnboundedFollowing))
+  }
+
+  /**
+   * Returns a new [[WindowFunctionDefinition]], with fixed number of records.
+   * For example:
+   * {{{
+   *   df.over(partitionBy("k1").orderBy($"k2", $"k3").rows)
+   * }}}
+   * @group window_funcs
+   */
+  def rows: WindowFunctionDefinition = {
+    new WindowFunctionDefinition(column, partitionSpec, orderSpec,
+      SpecifiedWindowFrame(RowFrame, UnboundedPreceding, UnboundedFollowing))
+  }
+
+  /**
+   * Returns a new [[WindowFunctionDefinition]], with position specified preceding of CURRENT_ROW.
+   * It can be either Lower or Upper Bound position depends on the semantic context.
+   * For example:
+   * {{{
+   *   // [CURRENT_ROW - 1, ~)
+   *   df.over(partitionBy("k1").orderBy("k2").rows.preceding(1))
+   *   // [CURRENT_ROW - 3, CURRENT_ROW - 1]
+   *   df.over(partitionBy("k1").orderBy("k2").rows.between.preceding(3).and.preceding(1))
+   *   // (~, CURRENT_ROW - 1]
+   *   df.over(partitionBy("k1").orderBy("k2").rows.between.unboundedPreceding.and.preceding(1))
+   * }}}
+   * @group window_funcs
+   */
+  def preceding(n: Int): WindowFunctionDefinition = {
+    assert(n > 0)
```

Comment:
Just confirm, Hive will take `0 preceding` as `CURRENT_ROW`, I will follow the same pattern